File: src/main/java/com/ptaf/ai/policy/AiPolicy.java
package com.ptaf.ai.policy;

import com.ptaf.ai.security.LogRedactor;
import org.yaml.snakeyaml.Yaml;

import java.io.InputStream;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.regex.Pattern;

/** Loads optional {@code /config/ai_policy.yml} for Phase 4 guardrails. */
public final class AiPolicy {

    private final Map<String, Object> root;

    public AiPolicy() {
        this(load());
    }

    AiPolicy(Map<String, Object> root) {
        this.root = root;
    }

    private static Map<String, Object> load() {
        try (InputStream in = AiPolicy.class.getResourceAsStream("/config/ai_policy.yml")) {
            if (in == null) {
                return Collections.emptyMap();
            }
            Yaml yaml = new Yaml();
            @SuppressWarnings("unchecked")
            Map<String, Object> map = yaml.load(in);
            return map != null ? map : Collections.emptyMap();
        } catch (Exception e) {
            return Collections.emptyMap();
        }
    }

    @SuppressWarnings("unchecked")
    private Map<String, Object> policy() {
        Object v = root.get("ai_policy");
        if (v instanceof Map<?, ?> m) {
            return (Map<String, Object>) m;
        }
        return Collections.emptyMap();
    }

    public int maxRequirementChars() {
        Object v = policy().get("max_requirement_chars");
        if (v instanceof Number n) {
            return n.intValue();
        }
        return 20_000;
    }

    public int duplicateStepWarnThreshold() {
        Object v = policy().get("duplicate_step_warn_threshold");
        if (v instanceof Number n) {
            return n.intValue();
        }
        return 8;
    }

    public int duplicateStepFailThreshold() {
        Object v = policy().get("duplicate_step_fail_threshold");
        if (v instanceof Number n) {
            return n.intValue();
        }
        return 25;
    }

    public boolean redactTriageInput() {
        Object v = policy().get("redact_triage_input");
        if (v instanceof Boolean b) {
            return b;
        }
        return true;
    }

    /** Apply policy to optional log paste before Gemini triage. */
    public String maybeRedactTriageInput(String text) {
        if (text == null) {
            return null;
        }
        return redactTriageInput() ? LogRedactor.redact(text) : text;
    }

    public List<Pattern> blockedRequirementPatterns() {
        Object v = policy().get("blocked_requirement_patterns");
        if (!(v instanceof List<?> list)) {
            return List.of();
        }
        List<Pattern> out = new ArrayList<>();
        for (Object o : list) {
            if (o != null) {
                try {
                    out.add(Pattern.compile(o.toString()));
                } catch (Exception ignored) {
                    // skip invalid
                }
            }
        }
        return out;
    }

    /** @return null if OK, or rejection reason */
    public String validateRequirement(String requirement) {
        if (requirement == null) {
            return "requirement is null";
        }
        int max = maxRequirementChars();
        if (max > 0 && requirement.length() > max) {
            return "requirement exceeds max_requirement_chars (" + max + ")";
        }
        for (Pattern p : blockedRequirementPatterns()) {
            if (p.matcher(requirement).find()) {
                return "requirement matched blocked pattern: " + p.pattern();
            }
        }
        return null;
    }
}

File: src/main/ptaf/ai/policy/ai_policy.py
from __future__ import annotations

import re
from pathlib import Path
from typing import Any

import yaml

from ptaf.ai.security.log_redactor import redact


class AiPolicy:
    def __init__(self, root: dict[str, Any] | None = None) -> None:
        self._root = root if root is not None else self._load()

    @staticmethod
    def _load() -> dict[str, Any]:
        path = Path(__file__).resolve().parents[5] / "src" / "test" / "resources" / "config" / "ai_policy.yml"
        if not path.is_file():
            return {}
        with path.open(encoding="utf-8") as handle:
            data = yaml.safe_load(handle)
        return data if isinstance(data, dict) else {}

    def _policy(self) -> dict[str, Any]:
        value = self._root.get("ai_policy")
        return value if isinstance(value, dict) else {}

    def max_requirement_chars(self) -> int:
        value = self._policy().get("max_requirement_chars")
        return int(value) if isinstance(value, (int, float)) else 20_000

    def duplicate_step_warn_threshold(self) -> int:
        value = self._policy().get("duplicate_step_warn_threshold")
        return int(value) if isinstance(value, (int, float)) else 8

    def duplicate_step_fail_threshold(self) -> int:
        value = self._policy().get("duplicate_step_fail_threshold")
        return int(value) if isinstance(value, (int, float)) else 25

    def redact_triage_input(self) -> bool:
        value = self._policy().get("redact_triage_input")
        return bool(value) if isinstance(value, bool) else True

    def maybe_redact_triage_input(self, text: str | None) -> str | None:
        if text is None:
            return None
        return redact(text) if self.redact_triage_input() else text

    def blocked_requirement_patterns(self) -> list[re.Pattern[str]]:
        value = self._policy().get("blocked_requirement_patterns")
        if not isinstance(value, list):
            return []
        patterns: list[re.Pattern[str]] = []
        for item in value:
            if item is None:
                continue
            try:
                patterns.append(re.compile(str(item)))
            except re.error:
                continue
        return patterns

    def validate_requirement(self, requirement: str | None) -> str | None:
        if requirement is None:
            return "requirement is null"
        max_chars = self.max_requirement_chars()
        if max_chars > 0 and len(requirement) > max_chars:
            return f"requirement exceeds max_requirement_chars ({max_chars})"
        for pattern in self.blocked_requirement_patterns():
            if pattern.search(requirement):
                return f"requirement matched blocked pattern: {pattern.pattern}"
        return None
